content/en-us/luau/type-checking.md

```diff
@@ -0,0 +1,326 @@
+---
+title: Luau Typing
+description: Luau uses gradual typing through the use
+of type annotations and inference.
+---
+
+Luau uses a typing engine which gives every variable an inferred type.
+
+```lua
+local foo = "bar" --> string
+local x = 5 --> number
+```
+
+## Defining a type
+A type can be defined using the `type` keyword
+
+```lua
+type Vector2 = {x: number, y: number}
+```
+
+## Inference Modes
+There are three type inference modes that Luau can use in `Class.Script|Scripts`
+
+* `--!nocheck` - Don't check types
+* `--!nonstrict` - Default mode for all scripts, only asserts variable types if they are explicitly annotated
+* `--!strict` - Asserts all types based off the inferred or explicitly annotated type
+
+## Types
+
+A type annoatation can be defined using the `:` operator after a variable, followed by a type definition, by default, in `nonstrict` mode, all variables are assigned the type `any`.
+
+```lua
+local foo: string = "bar"
+local x: number = 5
+```
+
+There are 4 primitive types that can be used in an annotation:
+* `nil` - no value
+* `boolean` - `true`/`false` value
+* `number` - a numeric value
+* `string` - text
+
+Functions and tables are defined differently, and within Roblox, all `Class.Instance|Instances` and `DataType.Enums` have their own types
+
+```lua
+part.Touched:Connect(function(otherPart: BasePart)
+	local brickColor: Enum.BrickColor = otherPart.BrickColor
+end)
+```
+
+An optional type can be defined using `?` at the end of the annotation.
+
+```lua
+local foo: string? = nil
+```
+
+## Function Typing
+
+Assume the following function
+```lua
+local function add(x, y)
+	return x + y
+end
+```
+
+This is a fairly simple function where the behavior is well defined. Take the value of `x` and add it to `y`. However, nothing is stopping someone from using a `string` type in the `add` function, which causes an error.
+
+This is because Luau doesn't know that we want to use numbers in this function, lets rework this function such that it does constrain `x` and `y` to be numbers.
+
+```lua
+local function add(x: number, y: number)
+	return x + y
+end
+```
+
+By adding type annotations to the parameters of the function, Luau now knows that the function takes two numbers, and throws a warning if you try to put anything into the function that isn't a number
+
+```lua
+add(5, 10) -- ok
+add(5, "foo") -- TypeError: Type `string` could not be converted into `number`.
+```
+
+To define a return type, put a `:` operator at the end of the function definition
+
+```lua
+local function add(x: number, y: number): number
+```
+
+To return multiple types, place all return types in brackets
+
+```lua
+local function FindSource(script: BaseScript, pattern: string): (string, number)
+```
+
+### Defining a functional type
+
+A functional type can be defined by using the syntax `(in) -> out`, for example, the type signatures of the functions above is

```

Comment:
```suggestion
A functional type can be defined by using the syntax `(in) -> out`. Using the functions from the previous examples, the type signatures of the functions are:
```

In the future, IX (or you, if you continue to further contribute to the typechecking article) may want to poke around a bit at this one. Since the Creator Docs are meant for general consumption and are presenting the docs from Luau Lang high-level for developers of all skill levels to understand, "type signatures" may not be an easily understood term here. It may also be difficult to localise and for international users to figure out what that means as this is the only place in the article where the term is used.